### PR TITLE
Publish an event when a user first accesses scheduled activities.

### PR DESCRIPTION
**Pull request by alxdarksage**
In addition, remove the "synthetic" events two_weeks_before_enrollment and two_months_before_enrollment because these are now implementable using automatic custom events, and in addition, they are not in use in any study right now (they were created for FHPS).

The event you can now schedule against is "activities_retrieved" and it's similar to enrollment in that it's immutable.

### CODE REVIEW

File: app/org/sagebionetworks/bridge/services/ActivityEventService.java [hunk 1]
```diff
@@ -157,4 +169,25 @@ public void deleteActivityEvents(String healthCode) {
...
+            }
+        }        
+    }
+    
+    private Tuple<String> parseAutoEventValue(String automaticEventValue) {
+        // though Period serialization does not contain semicolons, limit the split to the first occurrence
+        String[] valueElements = automaticEventValue.split(":", 2); 
+        return (valueElements.length == 1) ?
+            new Tuple<>(ActivityEventObjectType.ENROLLMENT.name().toLowerCase(), valueElements[0]) :
+            new Tuple<>(valueElements[0], valueElements[1]);
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/ActivityEventService.java]: Do we need to special case enrollment events? I thought we had decided to just move these wholesale to activities_retrieved. Enrollment only happens on consent, which means this is a less reliable way to create automatic custom events.

File: app/org/sagebionetworks/bridge/services/ScheduledActivityService.java [hunk 2]
```diff
@@ -192,10 +195,14 @@ protected DateTime getDateTime() {
         List<ScheduledActivity> saves = performMerge(scheduledActivities, dbMap);
         activityDao.saveActivities(saves);
         
-        return orderActivities(scheduledActivities, V3_FILTER);
+        List<ScheduledActivity> orderedActivities = orderActivities(scheduledActivities, V3_FILTER);
+        String healthCode = context.getCriteriaContext().getHealthCode();
+        activityEventService.publishActivitiesRetrieved(study, healthCode, DateUtils.getCurrentDateTime());
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/ScheduledActivityService.java]: We need to publish the event _before_ we schedule them, since the schedule depends on this event. (Same with V4.)

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

File: app/org/sagebionetworks/bridge/services/ActivityEventService.java (hunk 1 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/services/ActivityEventService.java]: This was to ensure that the current configuration in mPower 2.0 works as currently specified (no event in the string). You are right that this is the only use of it so we could probably assume it will be fixed (I updated the editor to handle this origin event specification yesterday, it's in staging).

However I should validate that the automatic custom events when you save a study (to ensure they are the new format), and I'm not doing that. I will add it.

I don't see any reason to entirely avoid triggering off of enrollment if that makes sense for a study (or some other future event if we think of one besides these two, or a custom event that the study designers create and add as part of their study, etc.).

Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/services/ActivityEventService.java]: Also realizing I need to parse this so that "custom:myEvent:P3W" works correctly (custom events will have two colons).

File: app/org/sagebionetworks/bridge/services/ScheduledActivityService.java (hunk 2 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/services/ScheduledActivityService.java]: Oh boy, very true! Moved.

File: app/org/sagebionetworks/bridge/BridgeUtils.java [hunk 3]
```diff
@@ -66,6 +68,15 @@
     // "request context" object into every method of every class.
     private static final ThreadLocal<String> REQUEST_ID_THREAD_LOCAL = ThreadLocal.withInitial(() -> null);
 
+    public static Tuple<String> parseAutoEventValue(String automaticEventValue) {
+        // Will property split something like "custom:myEvent:P3W"
+        int lastIndex = automaticEventValue.lastIndexOf(":P");
+        if (lastIndex == -1) {
+            return new Tuple<>(ActivityEventObjectType.ENROLLMENT.name().toLowerCase(), automaticEventValue); 
+        }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/BridgeUtils.java]: We shouldn't default to enrollment. mPower 2.0 configs are going to change anyway, so back-compat is a non-issue. Adding a special case for enrollment just adds complexity and confusion.

If we absolutely need a default for some reason, that default should be activities_retrieved.

File: app/org/sagebionetworks/bridge/services/ActivityEventService.java [hunk 4]
```diff
@@ -53,7 +54,11 @@ public void publishCustomEvent(Study study, String healthCode, String eventKey,
                 .withObjectType(ActivityEventObjectType.CUSTOM)
                 .withObjectId(eventKey)
                 .withTimestamp(timestamp).build();
-        activityEventDao.publishEvent(event);
+        
+        if (activityEventDao.publishEvent(event)) {
+            // Create automatic events, as defined in the study
+            createAutomaticCustomEvents(study, healthCode, event);
+        }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/ActivityEventService.java:61]: We have a circular dependency here. createAutomaticCustomEvents() calls publishCustomEvents() which calls createAutomaticCustomEvents(). It would be difficult to construct a custom event config that causes an infinite recursion here.

File: test/org/sagebionetworks/bridge/services/ActivityEventServiceTest.java [hunk 5]
```diff
@@ -225,6 +233,131 @@ public void canPublishEnrollmentEventWithAutomaticCustomEvents() {
...
+        
+        activityEventService.publishEnrollmentEvent(study,"AAA-BBB-CCC", new ConsentSignature.Builder().build());
+        
+        // Only happens once, none of the other custom events are published.
+        verify(activityEventDao, times(1)).publishEvent(any());
+    }
+    
+    @Test
+    public void whenNoActivitiesRetrievedEventPublishNoCustomEvents() {
+        // Configure study with automatic custom events
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/services/ActivityEventServiceTest.java]: Is it worth adding an equivalent test for custom events publishing automatic custom events?

File: app/org/sagebionetworks/bridge/validators/StudyValidator.java [hunk 6]
```diff
@@ -66,16 +73,39 @@ public void validate(Object obj, Errors errors) {
...
             if (study.getIdentifier().length() < 2) {
                 errors.rejectValue("identifier", "must be at least 2 characters");
             }
         }
         if (study.getActivityEventKeys().stream()
-                .anyMatch(k -> !k.matches(BridgeConstants.BRIDGE_IDENTIFIER_PATTERN))) {
-            errors.rejectValue("activityEventKeys", "must contain only lower-case letters and/or numbers with " +
-                    "optional dashes");
+                .anyMatch(k -> !k.matches(BridgeConstants.SYNAPSE_IDENTIFIER_PATTERN))) {
+            errors.rejectValue("activityEventKeys", SYNAPSE_IDENTIFIER_ERROR);
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/validators/StudyValidator.java]: Was this change intentional? Activity Event Keys have nothing to do with Synapse, so I'm not sure if the constants are correct.

File: app/org/sagebionetworks/bridge/validators/StudyValidator.java [hunk 7]
```diff
@@ -66,16 +73,39 @@ public void validate(Object obj, Errors errors) {
...
+        }
+        if (study.getAutomaticCustomEvents() != null) {
+            for (Map.Entry<String, String> entry : study.getAutomaticCustomEvents().entrySet()) {
+                String key = entry.getKey();
+                String value = entry.getValue();
+                
+                // Validate that the key follows the same rules for activity event keys
+                if (!key.matches(BridgeConstants.SYNAPSE_IDENTIFIER_PATTERN)) {
+                    errors.rejectValue("automaticCustomEvents["+key+"]", SYNAPSE_IDENTIFIER_ERROR);
+                }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/validators/StudyValidator.java]: Similarly, are we sure we're supposed to be using Synapse identifiers here?

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

File: app/org/sagebionetworks/bridge/BridgeUtils.java (hunk 3 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/BridgeUtils.java]: Okay we don't default any longer, and the validator will return validation errors if the string is something along the lines of "P3D". You need to add either a system event or a custom event key to trigger off of. The existing mPower 2.0 config won't be valid after deployment.

File: app/org/sagebionetworks/bridge/services/ActivityEventService.java (hunk 4 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/services/ActivityEventService.java:61]: Now calling the DAO directly rather than back into the service for the creation of these automatic custom events.

File: test/org/sagebionetworks/bridge/services/ActivityEventServiceTest.java (hunk 5 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/services/ActivityEventServiceTest.java]: Added similar tests that automatic custom events don't fire for enrollment or custom events, if the DAO returns false.

File: app/org/sagebionetworks/bridge/validators/StudyValidator.java (hunk 6 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/validators/StudyValidator.java]: Yes. Your current study burst keys are camel-cased and there's no reason they can't be camel-cased, this validation seemed too stringent. I wanted it to be possible for the custom activity event keys and the automatic event keys to have the same format constraints. The synapse validation allows for upper-case letters and numbers and seemed more appropriate for this.

File: app/org/sagebionetworks/bridge/validators/StudyValidator.java (hunk 7 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/validators/StudyValidator.java]: The regex seemed appropriate for event keys: [a-zA-Z0-9_-]. Will discuss on Slack if we want to create a different constant to avoid confusion.

**Review by DwayneJengSage — APPROVED** [review]
